packages/@aws-cdk/cdk/lib/core/tokens/token.ts

```diff
@@ -0,0 +1,130 @@
+ * mark values that need special processing at document rendering time.
+ *
+ * Tokens can be embedded into strings while retaining their original
+ * semantics.
+ */
+export class Token {
+  /**
+   * If this Token represents a reference, an identifier for the reference Type
+   */
+  public readonly referenceType?: string;
+
+  private tokenStringification?: string;
+  private tokenListification?: string[];
+
+  /**
+   * Creates a token that resolves to `value`.
+   *
+   * If value is a function, the function is evaluated upon resolution and
+   * the value it returns will be used as the token's value.
+   *
+   * displayName is used to represent the Token when it's embedded into a string; it
+   * will look something like this:
+   *
+   *    "embedded in a larger string is ${Token[DISPLAY_NAME.123]}"
+   *
+   * This value is used as a hint to humans what the meaning of the Token is,
+   * and does not have any effect on the evaluation.
+   *
+   * Must contain only alphanumeric and simple separator characters (_.:-).
+   *
+   * @param valueOrFunction What this token will evaluate to, literal or function.
+   * @param displayName A human-readable display hint for this Token
+   */
+  constructor(private readonly valueOrFunction?: any, private readonly displayName?: string) {
+  }
+
+  /**
+   * @returns The resolved value for this token.
+   */
+  public resolve(_context: ResolveContext): any {
+    let value = this.valueOrFunction;
+    if (typeof(value) === 'function') {
+      value = value();
+    }
+
+    return value;
+  }
+
+  /**
+   * Return a reversible string representation of this token
+   *
+   * If the Token is initialized with a literal, the stringified value of the
+   * literal is returned. Otherwise, a special quoted string representation
+   * of the Token is returned that can be embedded into other strings.
+   *
+   * Strings with quoted Tokens in them can be restored back into
+   * complex values with the Tokens restored by calling `resolve()`
+   * on the string.
+   */
+  public toString(): string {
+    const valueType = typeof this.valueOrFunction;
+    // Optimization: if we can immediately resolve this, don't bother
+    // registering a Token.
+    if (valueType === 'string' || valueType === 'number' || valueType === 'boolean') {
+      return this.valueOrFunction.toString();
+    }
+
+    if (this.tokenStringification === undefined) {
+      this.tokenStringification = TOKEN_MAP.registerString(this, this.displayName);
+    }
+    return this.tokenStringification;
+  }
+
+  /**
+   * Turn this Token into JSON
+   *
+   * This gets called by JSON.stringify(). We want to prohibit this, because
+   * it's not possible to do this properly, so we just throw an error here.
+   */
+  public toJSON(): any {
+    // tslint:disable-next-line:max-line-length
+    throw new Error('JSON.stringify() cannot be applied to structure with a Token in it. Use CloudFormationJSON.stringify() instead.');
```

Comment:
Now that we merged Token and CloudFormationToken, maybe we should just implement this here?

Reply:
That's not the same.

`CloudFormationJSON.stringify()` has to be called on the top-level structure that contains the tokens, while this is called for each Token individually as part of the `stringify()` operation.

Reply:
Right.